Fix `find_or_create_by`/`find_or_create_by!` when used within concurrent transactions

fatkodima opened the pull request:
Fixes https://github.com/rails/rails/issues/48035 (see there for the problem and investigation)

> Hum, should we only do that if we're inside a transaction?

Yes, this is not needed when outside of the transaction, so updated with that.

There is some nuance with this approach. For PostgreSQL ("read committed" isolation level by default), for example, previously the row read by `find_or_create_by` (when the `find_by` path of the whole method triggers) in the transaction acquired a read lock until the transaction ends, so other transactions can read/update this row. Now, this row is selected with `FOR UPDATE`, so no other transaction can modify it until the end of the transaction holding the lock. Seems like should be a rare problem. 

So, this change is really only needed for MySQL only (when in default isolation mode) or for any transactions with "repeatable read" mode, but it seems dirty to incorporate this knowledge into the `Relation` class.  

cc @byroot

fatkodima committed:
Fix `find_or_create_by`/`find_or_create_by!` when used within concurrent transactions

fatkodima commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -2452,3 +2452,75 @@ def custom_post_relation(alias_name = "omg_posts")
+      b_wakeup = Concurrent::Event.new
+
+      a = Thread.new do
+        Subscriber.transaction do
```
Here the "sqlite3 is busy" exceptions were raised - have no idea why, so made skipping this tests for sqlite. Seems like the same approach is used in https://github.com/rails/rails/blob/744b6715a5d07a2ddffceea8283f7b787541bc60/activerecord/test/cases/transactions_test.rb#L1377 (transactions and threads are also used there)

byroot replied:
>  have no idea why

It's because sqlite3-ruby doesn't release the GVL.

byroot commented:
Hum, the latest PG release is breaking CI, but other than that LGTM. I'll see about getting green main branch.

byroot commented:
@fatkodima CI is fixed, could you rebase just to be 100% sure?

fatkodima commented:
Rebased. I also refactored the duplication in tests, like is done in https://github.com/rails/rails/blob/611dac630555899fa75e7fa6901db38936fa5802/activerecord/test/cases/locking_test.rb#L793

jdelStrother commented:
Thanks both!
